corehq/apps/users/tasks.py

```diff
@@ -303,3 +306,33 @@ def update_domain_date(user_id, domain):
             and today > domain_membership.last_accessed):
         domain_membership.last_accessed = today
         user.save()
+
+
+@periodic_task(
+    run_every=deserialize_run_every_setting(settings.USER_REPORTING_METADATA_BATCH_SCHEDULE),
+    queue='background_queue',
+)
+def process_reporting_metadata_staging():
+    from corehq.apps.users.models import UserReportingMetadataStaging
+    from corehq.pillows.synclog import mark_last_synclog
+    from pillowtop.processors.form import mark_latest_submission
+
+    records = (
+        UserReportingMetadataStaging.objects.select_for_update(skip_locked=True).order_by('pk')
+    )[:100]
+    with transaction.atomic():
+        for record in records:
+            if record.received_on:
+                mark_latest_submission(
+                    record.domain, record.user_id, record.app_id, record.build_id,
+                    record.xform_version, record.form_meta, record.received_on
+                )
+            if record.device_id or record.sync_date:
+                mark_last_synclog(
+                    record.domain, record.user_id, record.app_id, record.build_id,
+                    record.sync_date, record.device_id
+                )
+            record.delete()
+
+    if UserReportingMetadataStaging.objects.exists():
+        process_reporting_metadata_staging.delay()
```

Comment:
This will keep a task processing these records until there are none left. This means that there could be multiple of these processes running. We may want to add harsher limits on how often this is getting run. Some thoughts are:

A) Add a localsetting so this won't get called passed a certain time
B) Add a parameter to the task so that the depth of these calls only goes X calls deep
C) Adding a check for the number of records. i.e. if only 5 were processed, don't call this again.

I don't think this work is necessary at the moment, but something I'm keeping in mind